### DISCUSSION TITLE
Handle constraints error messages defined in model

### DISCUSSION THREAD

**Discussion by sevdog** [encode/django-rest-framework#9348] [Ideas & Suggestions]
In [django 4.1](https://docs.djangoproject.com/en/4.1/ref/models/constraints/#violation-error-message) it was added to constraints a `violation_error_message` attribute which is used by django validation to provide an appropriate error message when the constraint is not respected.

Currently this field is not used by DRF and was not covered in #7438, which was created before this feature was introduced and therefore did not addressed it later.

It would be very useful to support using the `violation_error_message` as default `message` for `UniqueTogetherValidator` automatically created from model's constraints.

However there is an issue in this: because django uses `%` formatting while the DRF uses `.format` and this makes the two messages formats incompatible. Thus a solution should be found without introducing some kind of regression for "unique togheter" validation.

This feature is important because currently there is no easy way to customize the error message of a unique validator derived from a unique constraint.

**Comment by lovelydinosaur** [comment]
> because django uses % formatting while the DRF uses .format

Would you be able to add interlinking to the above so that we can compare & contrast?

(There was an attempt made to resolve this in #8863, which was reverted in #9330.)

**Comment by sevdog** [comment]
This is the "format" section in DRF `UniqueTogetherValidator`:

https://github.com/encode/django-rest-framework/blob/085b7e166ba80aa973645e5249b441f2dbdc0c96/rest_framework/validators.py#L98

https://github.com/encode/django-rest-framework/blob/085b7e166ba80aa973645e5249b441f2dbdc0c96/rest_framework/validators.py#L175-L178

This is the "format" section of django `BaseConstraint` (from which is derived the `UniqueConstraint`):

https://github.com/django/django/blob/d658a3162fbeb68d148d1b2fcf4da4fe1437eddb/django/db/models/constraints.py#L22

https://github.com/django/django/blob/d658a3162fbeb68d148d1b2fcf4da4fe1437eddb/django/db/models/constraints.py#L74-L75

There is also a big difference in which elements are used for formatting:

- DRF uses `field_names` to tell which fields are to be unique together
- Django uses `name` which is the DB constraint name (since it is also used by `CheckConstraint`)
